core/src/main/scala/org/apache/spark/BarrierCoordinator.scala

```diff
@@ -0,0 +1,229 @@
+      barrierEpochByStageIdAndAttempt.remove((stageInfo.stageId, stageInfo.attemptNumber))
+      cancelTimerTask(stageInfo.stageId, stageInfo.attemptNumber)
+    }
+  }
+
+  // Epoch counter for each barrier (stage, attempt).
+  private val barrierEpochByStageIdAndAttempt = new ConcurrentHashMap[(Int, Int), Int]
+
+  // Remember all the blocking global sync requests for each barrier (stage, attempt).
+  private val syncRequestsByStageIdAndAttempt =
+    new ConcurrentHashMap[(Int, Int), ArrayBuffer[RpcCallContext]]
+
+  // Remember all the TimerTasks for each barrier (stage, attempt).
+  private val timerTaskByStageIdAndAttempt = new ConcurrentHashMap[(Int, Int), TimerTask]
+
+  // Number of tasks for each stage.
+  private val numTasksByStage = new ConcurrentHashMap[Int, Int]
+
+  override def onStart(): Unit = {
+    super.onStart()
+    listenerBus.addToStatusQueue(listener)
+  }
+
+  /**
+   * Get the array of [[RpcCallContext]]s that correspond to a barrier sync request from a stage
+   * attempt.
+   */
+  private def getOrInitSyncRequests(
+      stageId: Int,
+      stageAttemptId: Int,
+      numTasks: Int): ArrayBuffer[RpcCallContext] = {
+    syncRequestsByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId),
+      new ArrayBuffer[RpcCallContext](numTasks))
+    syncRequestsByStageIdAndAttempt.get((stageId, stageAttemptId))
+  }
+
+  /**
+   * Clean up the array of [[RpcCallContext]]s that correspond to a barrier sync request from a
+   * stage attempt.
+   */
+  private def cleanupSyncRequests(stageId: Int, stageAttemptId: Int): Unit = {
+    val requests = syncRequestsByStageIdAndAttempt.remove((stageId, stageAttemptId))
+    if (requests != null) {
+      requests.clear()
+    }
+    logInfo(s"Removed all the pending barrier sync requests from Stage $stageId (Attempt " +
+      s"$stageAttemptId).")
+  }
+
+  /**
+   * Get the barrier epoch that correspond to a barrier sync request from a stage attempt.
+   */
+  private def getOrInitBarrierEpoch(stageId: Int, stageAttemptId: Int): Int = {
+    barrierEpochByStageIdAndAttempt.putIfAbsent((stageId, stageAttemptId), 0)
+    barrierEpochByStageIdAndAttempt.get((stageId, stageAttemptId))
+  }
+
+  /**
+   * Increase the barrier epoch that correspond to a barrier sync request from a stage attempt.
+   */
+  private def increaseBarrierEpoch(stageId: Int, stageAttemptId: Int): Unit = {
+    val barrierEpoch = barrierEpochByStageIdAndAttempt.getOrDefault((stageId, stageAttemptId), -1)
+    if (barrierEpoch >= 0) {
+      barrierEpochByStageIdAndAttempt.put((stageId, stageAttemptId), barrierEpoch + 1)
+    } else {
+      // The stage attempt already finished, don't update barrier epoch.
+    }
+  }
+
+  /**
+   * Cancel TimerTask for a stage attempt.
+   */
+  private def cancelTimerTask(stageId: Int, stageAttemptId: Int): Unit = {
+    val timerTask = timerTaskByStageIdAndAttempt.get((stageId, stageAttemptId))
+    if (timerTask != null) {
```

Comment:
Not safe. Use

~~~scala
val timerTask = timerTaskByStageIdAndAttempt.remove((stageId, stageAttemptId))
if (timerTask != null) {
  timerTask.cancel()
}
~~~